pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,353 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os/exec"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+var journal = journalServer{}
+
+// journalServer returns text output from the OS specific service logger to view
+// from the client. It runs with the privileges of the calling  process
+// (the kubelet) and should only be allowed to be invoked by a root user.
+type journalServer struct{}
+
+// ServeHTTP translates HTTP query parameters into arguments to be passed
+// to journalctl on the current system. It supports content-encoding of
+// gzip to reduce total content size.
+func (journalServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	var out io.Writer = w
+
+	args, errs := newJournalArgsFromURL(req.URL.Query())
+	if len(errs) > 0 {
+		http.Error(w, errs.ToAggregate().Error(), http.StatusBadRequest)
+		return
+	}
+
+	// TODO: Also set a response header that indicates how the request's query was resolved,
+	// e.g. "kube-log-source: journal://foobar?arg1=value" or "kube-log-source: file:///var/log/foobar.log"
+	w.Header().Set("Content-Type", "text/plain;charset=UTF-8")
+	if req.Header.Get("Accept-Encoding") == "gzip" {
+		w.Header().Set("Content-Encoding", "gzip")
+
+		gz := gzip.NewWriter(out)
+		defer gz.Close()
```

Comment:
the apiserver uses a pool for gzip writers and changes the default gzip encoding level for performance reasons... should we be doing the same here?

Reply:
https://github.com/kubernetes/kubernetes/blob/915924e64d08b7b2d3e6da81840e46d96304cdc9/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L140-L159

https://github.com/kubernetes/kubernetes/blob/915924e64d08b7b2d3e6da81840e46d96304cdc9/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L234-L235

https://github.com/kubernetes/kubernetes/blob/915924e64d08b7b2d3e6da81840e46d96304cdc9/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L247-L259

Reply:
We probably won't need Pool (can revisit) but can set the encoding level for the writer.